### ISSUE TITLE
`torch::train` step takes `devices` as a parameter

### ISSUE THREAD

**Issue by dirkgr** [allenai/tango#116]
Since `devices` is an input to `torch::train`, it'll be part of the step's unique id. That means it'll recompute when the devices change. I think it's safe to assume that training a model on 0 GPUs vs. 1 GPU vs. 2 GPUs will produce the same model, so this should not be part of the parameters. We should automatically use all GPUs we can take advantage of and use `CUDA_VISIBLE_DEVICES` to restrict their use.

**Comment by epwalsh** [comment]
> We should automatically use all GPUs we can take advantage of and use CUDA_VISIBLE_DEVICES to restrict their use.

I disagree with that. Moving up to multi-GPU training from single GPU training requires additional care to make sure your data gets properly sharded, and it also changes your effective batch size and the way instances are sampled. So training on multiple GPUs **will** likely produce a different model.

**Comment by dirkgr** [comment]
If I have a total batch of size 32, and I send 16 instances to 2 GPUs each, I should end up with the same model afterwards, no? It should work just like gradient accumulation, except fast.

**Comment by epwalsh** [comment]
> If I have a total batch of size 32, and I send 16 instances to 2 GPUs each, I should end up with the same model afterwards, no?

Ideally, yes, but in practice probably not. It might be tricky to set your RNG states correctly to guarantee you get the same results, and I think you'd also have differences with `LayerNorm`.

But my point about batch size is that we treat the `batch_size` parameter as the batch size per GPU.

**Comment by dirkgr** [comment]
Layer norm would be fine, but batch norm would not.

Fair enough. So if we take that to the logical conclusion, we have the opposite problem: `devices=None` will produce a different outcome with the same parameters, and thus with the same unique ID, depending on machine you run it on. I think we should switch it up so `devices` is required, but in the step's `__init__()` method we autodetect devices and fill in the parameter if we have to.

**Comment by epwalsh** [comment]
Actually I think the current behavior when `devices=None` is okay. When `devices=None` we train on a single GPU if one is available, otherwise we train on CPU. We never train on multiple GPUs unless you explicitly set `devices`. And it's probably reasonable to assume that you get the same model if you train on a single GPU or CPU only.
